src/3d/terrain/qgsquantizedmeshterraingenerator.cpp

```diff
@@ -0,0 +1,372 @@
+  else
+    clone->mLayerRef = mLayerRef; // Copy just the reference
+  return clone;
+}
+
+QgsTerrainGenerator::Type QgsQuantizedMeshTerrainGenerator::type() const
+{
+  return QgsTerrainGenerator::QuantizedMesh;
+}
+
+QgsRectangle QgsQuantizedMeshTerrainGenerator::rootChunkExtent() const
+{
+  return mMetadata->mBoundingVolume.bounds().toRectangle();
+}
+
+float QgsQuantizedMeshTerrainGenerator::rootChunkError( const Qgs3DMapSettings &map ) const
+{
+  Q_UNUSED( map );
+  return mMetadata->geometricErrorAtZoom( -1 );
+}
+
+void QgsQuantizedMeshTerrainGenerator::rootChunkHeightRange( float &hMin, float &hMax ) const
+{
+  hMin = mMetadata->mBoundingVolume.bounds().zMinimum();
+  hMax = mMetadata->mBoundingVolume.bounds().xMaximum();
+}
+float QgsQuantizedMeshTerrainGenerator::heightAt( double x, double y, const Qgs3DRenderContext &context ) const
+{
+  // TODO: This is the interesting part! We can read the height from the best
+  // currently loaded tile, or fetch the most precise tile for the coordinates
+  // given, but both have downsides.
+  Q_UNUSED( x );
+  Q_UNUSED( y );
+  Q_UNUSED( context );
+  return 0;
+}
+
+void QgsQuantizedMeshTerrainGenerator::writeXml( QDomElement &elem ) const
+{
+  QDomDocument doc = elem.ownerDocument();
+
+  elem.setAttribute( QStringLiteral( "layer" ), mLayerRef.layerId );
+}
+
+void QgsQuantizedMeshTerrainGenerator::readXml( const QDomElement &elem )
+{
+  QgsMapLayerRef layerRef = QgsMapLayerRef( elem.attribute( QStringLiteral( "layer" ) ) );
+  // We can't call setLayer yet, the reference is not resolved
+  mLayerRef = layerRef;
+}
+
+void QgsQuantizedMeshTerrainGenerator::resolveReferences( const QgsProject &project )
+{
+  mLayerRef.resolve( &project );
+  setLayer( layer() );
+}
+
+QgsChunkLoader *QgsQuantizedMeshTerrainGenerator::createChunkLoader( QgsChunkNode *node ) const
+{
+  long long tileId = QgsQuantizedMeshIndex::encodeTileId( nodeIdToTile( node->tileId() ) );
+  return new QgsQuantizedMeshTerrainChunkLoader( mTerrain, node, tileId, mIndex, mTileCrsToMapCrs );
+}
+
+QgsChunkNode *QgsQuantizedMeshTerrainGenerator::createRootNode() const
+{
+  return new QgsChunkNode(
+  {0, 0, 0},
+  mRootBbox, // Given to us by setupQuadtree()
+  mMetadata->geometricErrorAtZoom( -1 ) );
+}
+
+QVector<QgsChunkNode *> QgsQuantizedMeshTerrainGenerator::createChildren( QgsChunkNode *node ) const
+{
+  QVector<QgsChunkNode *> children;
+
+  for ( auto offset : std::vector<std::pair<int, int>> {{0, 0}, {0, 1}, {1, 0}, {1, 1}} )
+  {
+    QgsChunkNodeId childId(
+      node->tileId().d + 1,
+      node->tileId().x * 2 + offset.first,
+      node->tileId().y * 2 + offset.second
+    );
+    QgsTileXYZ tile = nodeIdToTile( childId );
+    if ( !mMetadata->containsTile( tile ) )
+      continue;
+
+    QgsTileMatrix zoomedTileMatrix = QgsTileMatrix::fromTileMatrix( tile.zoomLevel(), mMetadata->mTileMatrix );
+    QgsRectangle extent2d = mTileCrsToMapCrs.transform( zoomedTileMatrix.tileExtent( tile ) );
```

Comment:
transform may also throw QgsCsException

Reply:
Thanks for the review. Do you think I should log the error and skip that child?